File: scala/147.scala
object Solution {
    def insertionSortList(head: ListNode): ListNode = {
        val dummy = new ListNode(-5001)
        var ptr = head
        var e = dummy
        var prev = dummy
        while (ptr != null) {
          e = dummy
          prev = dummy
          while (ptr.x > e.x && e.next != null) {
            prev = e
            e = e.next
          }
          val n = ptr.next
          if (ptr.x > e.x) {
            e.next = ptr
            ptr.next = null
            ptr = n
          } else {
            ptr.next = e
            prev.next = ptr
            ptr = n
          }
        }
        dummy.next
    }
}

File: leetcodeJava/147.java
package leetcodeJava;

class Solution {
    public ListNode insertionSortList(ListNode head) {
        var dummyHead = new ListNode(-5001);
        ListNode tail = null;
        var ptr = head;
        while(ptr != null) {
            var next = ptr.next;
            add(dummyHead, ptr);
            ptr = next;
        }
        return dummyHead.next;
    }
    
    private void add(ListNode head, ListNode node) {
        var ptr = head;
        var prev = ptr;
        while (ptr != null) {
            if (ptr.val >= node.val) break;
            prev = ptr;
            ptr = ptr.next;
        }
        prev.next = node;
        node.next = ptr;
    }
}
